p2p/transport/webrtc/udpmux/packetbuffer.go

```diff
@@ -0,0 +1,95 @@
+package udpmux
+
+import (
+	"context"
+	"io"
+	"net"
+	"sync"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+type packet struct {
+	addr net.Addr
+	size int
+}
+
+type packetBuffer struct {
+	ctx         context.Context
+	mu          sync.Mutex
+	pdata       *pool.Buffer
+	pkts        []packet
+	notify      chan struct{}
+	readWaiting bool
+}
+
+func newPacketBuffer(ctx context.Context) *packetBuffer {
+	return &packetBuffer{
+		ctx:    ctx,
+		pdata:  new(pool.Buffer),
```

Comment:
This bit confuses me too. Looking at https://github.com/libp2p/go-buffer-pool/blob/a29bd39bcbb7ff3e385e270e26ff5686231b7860/buffer.go#L67 , it seems like a buffer from the pool will be fetched on read anyway. The downside here seems like it will be repeatedly grown as more packet data is backed up in the buffer. I'm not sure what you mean by global buffer. If I use `GlobalPool.Get`, it will return a byte slice. I can use it to initialize the `Buffer` object but that would be redundant. I also want to use the `Buffer` object because it can size the buffer using `grow` and `sink`.

Reply:
Have a look at how we used it here, for example: https://github.com/libp2p/go-libp2p/blob/4c74181d0dada0b4d600885285b96afee74e223d/core/record/envelope.go#L290

Remember you also have to return a buffer to the pool after using it.

Reply:
Wouldn't I have to manually grow and shrink this byte slice if the amount of data buffered exceeds the current buffer cap? The `Buffer` object also fetches its buffers from the global pool (at least in this case) https://github.com/libp2p/go-buffer-pool/blob/a29bd39bcbb7ff3e385e270e26ff5686231b7860/buffer.go#L97.